dotnet/sdk-phase-1.md

```diff
@@ -0,0 +1,1101 @@
+  * Leaning towards "Against"
+
+### HostBuilder and Worker Services
+
+https://learn.microsoft.com/en-us/dotnet/core/extensions/generic-host and
+https://learn.microsoft.com/en-us/dotnet/core/extensions/workers
+
+* Same discussion around dependencies as logging, this would need `Microsoft.Extensions.Hosting`
+* Our `Temporalio.Worker.TemporalWorker` is not a subclass of `BackgroundService` to avoid the dependency
+  * `TemporalWorker.ExecuteAsync` matches `BackgroundService.ExecuteAsync` very intentionally, so this is incredibly
+    easy for a user to create and we will of course have a sample
+
+## General Decision Justifications/Discussion
+
+#### Why method references instead of proxy lambdas
+
+Some utilities like Azure Durable Entities support invoking a method on a proxied version of an item instead of just
+referencing the method. There are pros and cons. To discuss, here are examples and cons of each approach:
+
+##### With method references
+
+```csharp
+// Start a workflow
+var handle = await client.StartWorkflowAsync(
+    IMyWorkflow.Ref.RunAsync, "some arg", new(ID: "my-workflow-id", TaskQueue: "my-task-queue"));
+
+// Send a signal
+await handle.Signal(
+    IMyWorkflow.Ref.SignalAsync, "some arg");
+
+// Invoke activity from workflow (future guess)
+await ExecuteActivity(
+    MyActivities.Ref.Activity, "some arg", new(StartToCloseTimeout: TimeSpan.FromHours(5)));
+```
+
+Cons:
+
+* You have to make a reference to the class whose methods you need to reference
+* You can't have the handle typed with the workflow class
+  * So for example you won't get a compile-time failure if you invoke a signal from some other workflow
+  * With lambdas we might be able to have them typed though
+* Can only support 0 or 1 parameters
+  * Can technically add generic overloads for more parameters, but you have to stop somewhere
+  * This is an inherent problem in .NET in several things and why you often see a dozen overloads for a method
+
+##### With lambda
+
+```csharp
+// Start a workflow
+var handle = await client.StartWorkflowAsync<IMyWorkflow>(
+    myWf => myWf.RunAsync("some arg"), new(ID: "my-workflow-id", TaskQueue: "my-task-queue"));
+
+// Send a signal
+await handle.Signal(
+    myWf => myWf.SignalAsync("some arg"));
+
+// Invoke activity from workflow (future guess)
+await ExecuteActivity<MyActivities>(
+    myAct => myAct.Activity("some arg"), new(StartToCloseTimeout: TimeSpan.FromHours(5)));
+```
+
+Cons:
+
+* Every method must be virtual or on an interface
+  * Cannot proxy regular methods which is a bit off-putting. This works for Azure because they make you opt-in to the
+    typed approach
+    [by forcing interfaces](https://learn.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-dotnet-entities#accessing-entities-through-interfaces) which we don't want to do
+  * Cannot proxy static methods
+    * We should support static activities, so you'd have method references anyways probably
```

Comment:
Forgive my ignorance, but what does proxy mean in this context? I thought all Temporal might need to do with these functions is run something before and/or after innovation, is there something more than that?

Reply:
> Forgive my ignorance, but what does proxy mean in this context?

http://www.castleproject.org/projects/dynamicproxy/

> I thought all Temporal might need to do with these functions is run something before and/or after innovation, is there something more than that?

You need to be able to reference the function. When you have a client and need to call a workflow/signal/query, or in a workflow and need to call an activity/child, you need a typed reference to them for type safety. C# doesn't have method references, so we need to reference them somehow.

> The lambda sample here looks far more idiomatic to c#

But did you know that lambda is not invoking the thing? Instead it's just using a lambda to be able to provide a method reference for use by what does schedule the work.

Here's the problem: I have a workflow/activity instance method, and I need to reference it (not call it) via a start workflow or start activity call. What is the best way to do that in C#?